### PR TITLE
Extract pod garbage collection into a separate controller

### PR DESCRIPTION
**Pull request by jveski**
Currently the synthesis lifecycle controller creates and deletes pods. But since it reconciles compositions, it needs to reconcile the entire set of pods associated with the composition for every reconciliation. This works, but it's confusing to read/test, and it makes requeing to check timeouts tricky.

This PR refactors the deletion logic into a separate controller, extends test coverage around some important edge cases, and simplifies the existing lifecycle controller somewhat. Mostly it unblocks better testing of the lifecycle controller in the future.

### CODE REVIEW

File: internal/controllers/synthesis/integration_test.go [hunk 1]
```diff
@@ -238,3 +242,38 @@ func TestControllerSwitchingSynthesizers(t *testing.T) {
 		assert.NotEqual(t, comp.Status.CurrentSynthesis.ResourceSlices, initialSlices)
 	})
 }
+
+func TestControllerBackoff(t *testing.T) {
```
Comment (jveski) [inline internal/controllers/synthesis/integration_test.go:246]: Unrelated to the PR: moved from backoff_test.go since it was the only test in the file

File: internal/controllers/synthesis/lifecycle.go [hunk 2]
```diff
@@ -197,6 +150,41 @@ func (c *podLifecycleController) Reconcile(ctx context.Context, req ctrl.Request
...
+		}
+
+		if podIsCurrent(comp, &pod) {
+			if !active {
+				active = true
+				continue
+			}
+
+			// Clean up any dupes
+			err := c.client.Delete(ctx, &pod)
```
Comment (jveski) [inline internal/controllers/synthesis/lifecycle.go:176]: There is one deletion case that can't logically be addressed by the podGC controller

File: internal/controllers/synthesis/lifecycle.go [hunk 3]
```diff
@@ -197,6 +166,41 @@ func (c *podLifecycleController) Reconcile(ctx context.Context, req ctrl.Request
...
+
+	var active bool
+	var deleting bool
+	for _, pod := range pods.Items {
+		if pod.DeletionTimestamp != nil {
+			if !deleting {
+				deleting = true
+				continue
+			}
+			return false, nil // don't create new pod if two are deleting
```
Comment (AYM1607) [inline internal/controllers/synthesis/lifecycle.go:182]: Mind expanding on why we adopt this heuristic?

File: internal/controllers/synthesis/lifecycle.go (hunk 2 above)
Comment (AYM1607) [reply internal/controllers/synthesis/lifecycle.go:192]: It feels weird to have this here, why can't it be done in the gc controller?

File: internal/controllers/synthesis/gc.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		logger = logger.WithValues("reason", "SynthesizerDeleted")
+		return ctrl.Result{}, p.deletePod(ctx, pod, logger)
+	}
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting synthesizer: %w", err)
+	}
+
+	// Pods from old syntheses
+	if !podIsCurrent(comp, pod) {
+		const gracePeriod = 250 * time.Millisecond
```
Comment (AYM1607) [inline internal/controllers/synthesis/gc.go:88]: Could informer drift be larger than this in underlays?

File: internal/controllers/synthesis/lifecycle.go (hunk 2 above)
Comment (jveski) [reply internal/controllers/synthesis/lifecycle.go:192]: Hmm let me see what the tradeoffs would be

File: internal/controllers/synthesis/lifecycle.go [hunk 5]
```diff
@@ -83,64 +79,37 @@ func (c *podLifecycleController) Reconcile(ctx context.Context, req ctrl.Request
...
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing pods: %w", err)
 	}
 
-	// Synthesis isn't possible without a synth
-	if syn == nil {
-		return ctrl.Result{}, nil
+	ctx = logr.NewContext(ctx, logger)
+	ok, err := c.safeToCreatePod(ctx, pods, comp)
```
[the remark this answers is not in the document]
Comment (jveski) [reply internal/controllers/synthesis/lifecycle.go:106]: Exactly - we check the backoff interval first to avoid making the apiserver call before the next retry is ready

File: internal/controllers/synthesis/gc.go (hunk 4 above)
Comment (jveski) [reply internal/controllers/synthesis/gc.go:88]: Let me see if there's a better way to handle this, or at least add some comments

File: internal/controllers/synthesis/lifecycle.go (hunk 3 above)
Comment (jveski) [reply internal/controllers/synthesis/lifecycle.go:182]: Will try to refactor a bit for clarity here

File: internal/controllers/synthesis/gc.go (hunk 4 above)
Comment (jveski) [reply internal/controllers/synthesis/gc.go]: Up'd it to 1 second just in case

Comment (jveski) [reply internal/controllers/synthesis/gc.go]: My take is that it's extremely unlikely informers will be behind by >1s and even if they are, recreating the pod isn't a terrible failure mode.

File: internal/controllers/synthesis/lifecycle.go (hunk 2 above)
Comment (jveski) [reply internal/controllers/synthesis/lifecycle.go]: Remove it! Turns out we don't really need it anyway. I reworked a few things to avoid even needing a pod index.

**Review by AYM1607 — APPROVED** [review]
LGTM
